File: go/pkg/config/merger.go
package config

// DeepMerge merges source into destination. 
// If a key exists in both and both are maps, it merges them recursively.
// If a key exists in both and the source is not a map, the source value overwrites the destination.
func DeepMerge(dst, src map[string]interface{}) map[string]interface{} {
	for k, v := range src {
		if v == nil {
			continue
		}

		// If value is a map, check if destination also has a map
		srcMap, isSrcMap := v.(map[string]interface{})
		dstVal, isDstExist := dst[k]
		dstMap, isDstMap := dstVal.(map[string]interface{})

		if isSrcMap && isDstExist && isDstMap {
			// Recursive merge
			dst[k] = DeepMerge(dstMap, srcMap)
		} else {
			// Hard overwrite or new key
			dst[k] = v
		}
	}
	return dst
}

File: rust/src/config/merger.rs
// -----------------------------------------------------------------------------
// ESSENTIAL PROCESS:
// Deep merges configuration maps and environment variables with precedence rules.
//
// DATA FLOW:
// Source Map + Target Map -> Deep Merge -> Unified Configuration Map
//
// KEY PARAMETERS:
// - dst: Destination map receiving merged keys.
// - src: Source map providing override keys.
// -----------------------------------------------------------------------------

use serde_yml::Value;

/// Recursive deep merge of src into dst.
/// 
/// Behavior matches Go implementation:
/// - If both are mappings, recurse.
/// - Otherwise, overwrite dst with src.
pub fn deep_merge(dst: &mut Value, src: &Value) {
    if let (Some(dst_map), Some(src_map)) = (dst.as_mapping_mut(), src.as_mapping()) {
        for (k, v) in src_map {
            if !dst_map.contains_key(k) || !v.is_mapping() {
                dst_map.insert(k.clone(), v.clone());
            } else {
                deep_merge(dst_map.get_mut(k).unwrap(), v);
            }
        }
    } else {
        *dst = src.clone();
    }
}
